File: client/typescript/test/pico/bench.mjs
import { connect } from '../../dist/index.js'

const ENDPOINT = process.env.PICO_ENDPOINT ?? 'http://127.0.0.1:4437'
const CT = 'application/octet-stream'

const pico = connect('pico', ENDPOINT)
const base = `/bench-${Date.now().toString(36)}`

function payload(size, tag) {
  const body = new Uint8Array(size)
  body.set(new TextEncoder().encode(tag))
  return body
}

function fmt(n) {
  return n >= 1000 ? `${(n / 1000).toFixed(1)}k` : n.toFixed(0)
}

async function timed(label, fn) {
  const started = performance.now()
  const out = await fn()
  const seconds = (performance.now() - started) / 1000
  return { label, seconds, ...out }
}

function report({ label, seconds, records, bytes }) {
  const rps = records / seconds
  const mbps = bytes / seconds / (1024 * 1024)
  console.log(
    `${label.padEnd(46)} ${String(records).padStart(7)} rec  ${seconds.toFixed(2).padStart(6)}s  ${fmt(rps).padStart(8)} rec/s  ${mbps.toFixed(1).padStart(7)} MB/s`,
  )
}

async function sequentialAppends(count, size) {
  const stream = pico.stream(`${base}/seq`)
  await stream.create(CT)
  const body = payload(size, 'seq')
  return timed(`sequential append (${size}B, 1 rec/req)`, async () => {
    for (let i = 0; i < count; i++) {
      await stream.append([body])
    }
    return { records: count, bytes: count * size }
  })
}

async function producerRun(name, count, size, config) {
  const stream = pico.stream(`${base}/${name}`)
  await stream.create(CT)
  const producer = stream.producer(`bench-${name}`, config)
  const body = payload(size, name)
  const result = await timed(
    `producer ${name} (linger ${config.lingerMs}ms, inflight ${config.maxInflight})`,
    async () => {
      const pendings = new Array(count)
      for (let i = 0; i < count; i++) {
        pendings[i] = await producer.send(body)
      }
      const seqs = await Promise.all(pendings.map((p) => p.durable()))
      await producer.close()
      for (let i = 0; i < count; i++) {
        if (seqs[i] !== i) throw new Error(`ordering violated at ${i}: got ${seqs[i]}`)
      }
      return { records: count, bytes: count * size }
    },
  )
  const head = await stream.head()
  if (head.next !== String(count)) throw new Error(`expected next ${count}, got ${head.next}`)
  return result
}

async function readBack(name, expected, size) {
  const stream = pico.stream(`${base}/${name}`)
  return timed(`read back ${name} (pages of 1000)`, async () => {
    let records = 0
    let bytes = 0
    for await (const record of stream.records('0', { batch: { count: 1000 } })) {
      records += 1
      bytes += record.body.byteLength
    }
    if (records !== expected) throw new Error(`expected ${expected} records, read ${records}`)
    return { records, bytes: bytes || records * size }
  })
}

async function main() {
  console.log(`endpoint ${ENDPOINT}`)
  console.log('')

  report(await sequentialAppends(300, 128))
  report(await producerRun('batch-1', 5000, 128, { lingerMs: 5, maxInflight: 1 }))
  report(await producerRun('batch-8', 5000, 128, { lingerMs: 5, maxInflight: 8 }))
  report(await producerRun('large-8', 1000, 8192, { lingerMs: 5, maxInflight: 8 }))
  report(await readBack('batch-8', 5000, 128))
  report(await readBack('large-8', 1000, 8192))

  const listing = await pico.list(`${base}/`, 100)
  for (const info of listing.streams) {
    await pico.delete(info.name)
  }
  console.log('')
  console.log(`cleaned up ${listing.streams.length} bench streams`)
}

main().catch((error) => {
  console.error(error)
  process.exit(1)
})

File: client/go/test/pico/bench.go
package main

import (
	"context"
	"errors"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
	"time"

	picomq "github.com/PicoMQ/picomq/client/go"
)

const contentType = "application/octet-stream"

type result struct {
	label   string
	elapsed time.Duration
	records int
	bytes   int64
}

func main() {
	endpoint := flag.String("endpoint", envOr("PICO_ENDPOINT", "http://127.0.0.1:4437"), "Pico endpoint")
	sequentialCount := flag.Int("sequential", 300, "number of sequential appends")
	batchCount := flag.Int("batch", 5000, "number of small producer records")
	largeCount := flag.Int("large", 1000, "number of large producer records")
	flag.Parse()

	ctx := context.Background()
	client, err := picomq.NewPico(*endpoint)
	if err != nil {
		log.Fatal(err)
	}
	defer client.CloseIdleConnections()

	base := "/bench-" + strconv.FormatInt(time.Now().UnixNano(), 36)
	defer cleanup(client, base)

	fmt.Printf("endpoint %s\n\n", *endpoint)
	run(sequentialAppends(ctx, client, base+"/seq", *sequentialCount, 128))
	run(producerRun(ctx, client, base+"/batch", "batch", *batchCount, 128))
	run(producerRun(ctx, client, base+"/large", "large", *largeCount, 8192))
	run(readBack(ctx, client, base+"/batch", "batch", *batchCount))
	run(readBack(ctx, client, base+"/large", "large", *largeCount))
}

func sequentialAppends(ctx context.Context, client *picomq.PicoClient, name string, count, size int) (result, error) {
	stream := client.Stream(name)
	if _, err := stream.Create(ctx, contentType, 0); err != nil {
		return result{}, err
	}
	body := payload(size, "seq")
	started := time.Now()
	for i := 0; i < count; i++ {
		if _, err := stream.Append(ctx, picomq.AppendRecord{Body: body}); err != nil {
			return result{}, err
		}
	}
	return result{label: fmt.Sprintf("sequential append (%dB, 1 rec/req)", size), elapsed: time.Since(started), records: count, bytes: int64(count * size)}, nil
}

func producerRun(ctx context.Context, client *picomq.PicoClient, name, label string, count, size int) (result, error) {
	stream := client.Stream(name)
	if _, err := stream.Create(ctx, contentType, 0); err != nil {
		return result{}, err
	}
	config := picomq.DefaultProducerConfig()
	producer := stream.NewProducer("bench-"+label, &config)
	defer producer.Close(context.Background())
	body := payload(size, label)
	started := time.Now()
	pendings := make([]*picomq.Pending, count)
	for i := 0; i < count; i++ {
		pending, err := producer.Send(ctx, picomq.AppendRecord{Body: body})
		if err != nil {
			return result{}, err
		}
		pendings[i] = pending
	}
	for i, pending := range pendings {
		seq, err := pending.Await(ctx)
		if err != nil {
			return result{}, err
		}
		if seq != uint64(i) {
			return result{}, fmt.Errorf("ordering violated at %d: got %d", i, seq)
		}
	}
	if err := producer.Close(ctx); err != nil {
		return result{}, err
	}
	elapsed := time.Since(started)
	info, err := stream.Head(ctx)
	if err != nil {
		return result{}, err
	}
	if info == nil || info.Next != strconv.Itoa(count) {
		return result{}, fmt.Errorf("expected next %d, got %v", count, info)
	}
	return result{label: fmt.Sprintf("producer %s (linger %s)", label, config.Linger), elapsed: elapsed, records: count, bytes: int64(count * size)}, nil
}

func readBack(ctx context.Context, client *picomq.PicoClient, name, label string, expected int) (result, error) {
	iterator := client.Stream(name).Records(picomq.RecordsOptions{From: client.Beginning(), Limits: picomq.ReadLimits{Count: 1000}})
	started := time.Now()
	var records int
	var bytes int64
	for {
		record, err := iterator.Next(ctx)
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			return result{}, err
		}
		records++
		bytes += int64(len(record.Body))
	}
	if records != expected {
		return result{}, fmt.Errorf("expected %d records, read %d", expected, records)
	}
	return result{label: fmt.Sprintf("read back %s (pages of 1000)", label), elapsed: time.Since(started), records: records, bytes: bytes}, nil
}

func payload(size int, tag string) []byte {
	body := make([]byte, size)
	copy(body, tag)
	return body
}

func run(value result, err error) {
	if err != nil {
		log.Fatal(err)
	}
	seconds := value.elapsed.Seconds()
	fmt.Printf("%-46s %7d rec  %6.2fs  %8s rec/s  %7.1f MB/s\n", value.label, value.records, seconds, formatRate(float64(value.records)/seconds), float64(value.bytes)/seconds/(1024*1024))
}

func formatRate(value float64) string {
	if value >= 1000 {
		return fmt.Sprintf("%.1fk", value/1000)
	}
	return fmt.Sprintf("%.0f", value)
}

func cleanup(client *picomq.PicoClient, base string) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	listing, err := client.List(ctx, base+"/", 100)
	if err != nil {
		log.Printf("cleanup: %v", err)
		return
	}
	for _, info := range listing.Streams {
		if _, err := client.Delete(ctx, info.Name); err != nil {
			log.Printf("cleanup %s: %v", info.Name, err)
		}
	}
	fmt.Printf("\ncleaned up %d bench streams\n", len(listing.Streams))
}

func envOr(name, fallback string) string {
	if value := os.Getenv(name); value != "" {
		return value
	}
	return fallback
}
